curp/src/server/storage/wal/pipeline.rs

```diff
@@ -0,0 +1,169 @@
+use std::{
+    io,
+    path::{Path, PathBuf},
+    task::Poll,
+};
+
+use clippy_utilities::OverflowArithmetic;
+use event_listener::Event;
+use flume::r#async::RecvStream;
+use futures::{FutureExt, StreamExt};
+use thiserror::Error;
+use tokio::task::JoinHandle;
+use tokio_stream::Stream;
+use tracing::error;
+
+use super::util::LockedFile;
+
+/// The temp file extension
+const TEMP_FILE_EXT: &str = ".tmp";
+
+/// The file pipeline, used for pipelining the creation of temp file
+pub(super) struct FilePipeline {
+    /// The directory where the temp files are created
+    dir: PathBuf,
+    /// The size of the temp file
+    file_size: u64,
+    /// The file receive stream
+    file_stream: RecvStream<'static, LockedFile>,
+    /// The stop event listener
+    stop_event: Event,
+    /// The handle of the background file creation task
+    handle: JoinHandle<io::Result<()>>,
+}
+
+impl FilePipeline {
+    /// Creates a new `FilePipeline`
+    #[allow(clippy::integer_arithmetic)] // Introduced by tokio::select! macro
+    pub(super) fn new(dir: PathBuf, file_size: u64) -> io::Result<Self> {
+        Self::clean_up(&dir)?;
+
+        let (file_tx, file_rx) = flume::bounded(1);
+        let stop_event = Event::new();
+        let mut stop_listener = stop_event.listen();
+        let dir_c = dir.clone();
+
+        let handle = tokio::spawn(async move {
+            let mut file_count = 0;
+            loop {
+                let file = Self::alloc(&dir_c, file_size, &mut file_count)?;
```

Comment:
I think it's better to pass `usize` directly here and then increment `file_count` inside the loop

Reply:
done